common/cloud_builder.py

```diff
@@ -0,0 +1,258 @@
+        service_account_file)
+
+  def _upload_to_gcs(self, local_file_path: str) -> str:
+    """Uploads a file to Google Cloud Storage."""
+    dest_file_name = os.path.basename(local_file_path)
+    self.bucket.blob(dest_file_name).upload_from_filename(local_file_path)
+    bucket_file_url = f'gs://{self.bucket_name}/{dest_file_name}'
+    logging.info('Uploaded %s to %s', local_file_path, bucket_file_url)
+    return bucket_file_url
+
+  def _get_current_commit_id(self) -> str:
+    """Gets the current commit ID from the local git repository."""
+    try:
+      commit_id = subprocess.check_output(['git', 'rev-parse',
+                                           'HEAD']).decode('utf-8').strip()
+      return commit_id
+    except subprocess.CalledProcessError as e:
+      logging.error('Error fetching the current commit ID: %s', e)
+      return ''
+
+  def _request_cloud_build(self, agent_pickle_url: str, results_pickle_url: str,
+                           new_result_filename: str) -> str:
+    """Requests Cloud Build to execute the operation."""
+    cloud_build_config = {
+        'steps': [
+            # Step 1: Download the pickle files from GCS bucket.
+            {
+                'name': 'bash',
+                'dir': '/workspace',
+                'args': ['-c', 'mkdir -p pickles']
+            },
+            {
+                'name': 'gcr.io/cloud-builders/gsutil',
+                'dir': '/workspace',
+                'args': ['cp', agent_pickle_url, 'pickles/agent.pkl']
+            },
+            {
+                'name': 'gcr.io/cloud-builders/gsutil',
+                'dir': '/workspace',
+                'args': [
+                    'cp', results_pickle_url, 'pickles/result_history.pkl'
+                ]
+            },
+            # Step 2: Prepare git repos.
+            {
+                'name':
+                    'gcr.io/cloud-builders/git',
+                'entrypoint':
+                    'bash',
+                'dir':
+                    '/workspace',
+                'args': [
+                    '-c',
+                    (f'git clone --no-checkout {OFG_REPO} ofg && cd ofg &&'
+                     'git fetch --all && '
+                     f'git checkout {self._get_current_commit_id()}')
+                ]
+            },
+            {
+                'name': 'gcr.io/cloud-builders/git',
+                'entrypoint': 'bash',
+                'dir': '/workspace',
+                'args': ['-c', f'git clone --depth=1 {OF_REPO} ofg/oss-fuzz']
+            },
+            # Step 3: Run the Python script with the pickle files
+            {
+                'name':
+                    'gcr.io/cloud-builders/docker',
+                'entrypoint':
+                    'bash',
+                'dir':
+                    '/workspace/ofg',
+                'args': [
+                    '-c',
+                    """
+                    # Install Python 3.11 and pip
```

Comment:
This is very brittle, and will break if `gcr.io/cloud-builders/docker` updates. Can we create our own Docker image for this with all of these dependencies preinstalled?

Reply:
Done:
1. Using a separate image.
2. Moved the `Dockerfile` content to a file.
3. Manually pushed the image to our ofg artifact registry.
4. Cloud build now pulls from the registry.